### ISSUE TITLE
Update themes for GTK+ 3.20

### ISSUE THREAD

**Issue by andrew160** [BunsenLabs/bunsen-themes#13]
Hi,

First of all, huge thanks for creating these themes! I've yet to find a prettier one :smiley:

Sadly after the GTK+ 3.20 update almost everything broke. I really hope it can be updated to be compatible again.

Thanks in advance :smiley:

**Comment by ghost** [comment]
That's a difficult question. I'll outline some of the difficulties for @BunsenLabs/artwork-team so we can get some feedback and a plan for this, if any.

The themes were not written by us from the ground up but are based on third-party themes, such as XFCE Greybird.
- It is important to note that these themes are made for gtk 3.16 which is in Debian Jessie. Changes accomodating gtk 3.20 would very much break backwards compatibility (seeing as it isn't even forward-compatible), so updating the current package is out of the question. We would need to create a forked packgage for gtk 3.20.
- If the theme the theme in question is based on  itself hasn't been ported to gtk3 yet, a port probably won't happen because rewriting 5+ themes from the ground up is extremely time consuming and a trial-and-error process. The lack of documentation and a standardized process on the GTK3.20 side doesn't help. Maybe some of us are willing to take this upon themselves though?
- We could rebase the themes on a 3.20-compatible theme package with huge developer backing in order to keep things up and running for 3.20 and later. Numix, Arc, and Vortex come to mind. Extremely interesting because probably most similar to the current Greybird version are also the **Mate desktop themes** at https://github.com/mate-desktop/mate-themes.
- If at all, we should probably start with porting at most two themes to gtk3.20. I'd even say it 's better to start with one.

**Comment by andrew160** [comment]
> It is important to note that these themes are made for gtk 3.16 which is in Debian Jessie. Changes accomodating gtk 3.20 would very much break backwards compatibility (seeing as it isn't even forward-compatible), so updating the current package is out of the question. We would need to create a forked packgage for gtk 3.20.

Ouch.. that's not good.
It should be noted I'm using these themes on Arch currently, so no need to rush as it'll probably be a while until Debian updates to 3.20.

> If the theme the theme in question is based on itself hasn't been ported to gtk3 yet, a port probably won't happen because rewriting 5+ themes from the ground up is extremely time consuming and a trial-and-error process.

That's understandable.

> The lack of documentation and a standardized process on the GTK3.20 side doesn't help.

Agreed. This has really bothered me when I wanted to edit themes.

> We could rebase the themes on a 3.20-compatible theme package with huge developer backing in order to keep things up and running for 3.20 and later. Numix, Arc, and Vortex come to mind. Extremely interesting because probably most similar to the current Greybird version are also the Mate desktop themes at https://github.com/mate-desktop/mate-themes.

This seems like the best option. I think it's quite likely Greybird will be updated (if Shimmer Project doesn't take on the task, Xubuntu probably will.)
But if not, the Mate themes are a great option like you said. It's getting tons of development lately.

Also, the Blackbird theme recently got updated for 3.20:
https://github.com/shimmerproject/Blackbird
So that's another option, maybe slightly easier than the Mate ones.

And yes, porting one of them is a good start. I'd be willing to help port the standard Bunsen theme.

Anyway, it seems the best thing to do now is just to wait it out?

**Comment by hhhorb** [comment]
We support Debian stable, so I'd vote that we leave any upgraded themes to the "unofficial" category.

Is there a Debian package for 3.2 yet? If so (link, please), I will gladly set up a sandbox and work on a compatible light and dark theme, the users can customize it further from there.

Thanks for bringing up this issue, cheers!

**Comment by ghost** [comment]
I'd rather not cheer because this topic is annoying as hell...anyway, gtk 3.20 seems currently to be in experimental https://packages.debian.org/search?keywords=libgtk-3-0&searchon=names&suite=all&section=all , and let me tell you, rightfully so, because as always they didn't bother to check for bugs and do proper QA before releasing. On Arch, there's a new minor version of gtk3.20 every other day (I've downgraded to 3.18 and will stay there for a long time).

IMO we should wait a month or two until things got back to normal, and until all themes we could rebase on have fixed themselves.

**Comment by Hjdskes** [comment]
Possibly relevant to those attempting to port:

https://www.reddit.com/r/gnome/comments/4erp9v/where_can_i_find_the_css_reference_for_gnome/

**Comment by hhhorb** [comment]
> I'd rather not cheer because this topic is annoying as hell

Lol, right!

**Comment by ju1ius** [comment]
@2ion Hi, long time crunchbang user here, now on custom Sid+Openbox+Bunsen
If the themes are rarely updated, there should be a strict version constraint against libgtk-3.0 in the package depends. This way those of us running sid/testing don't find themselves with a clunky desktop after an apt-get upgrade.:wink:

**Comment by johnraff** [comment]
@2ion `Depends: libgtk-3-0 (<< 3.20)` would imply that  the themes will work with 3.18.9-1 for example. That's only available in Debian Sid for _powerpcspe_ architecture, but are we OK saying that any version less than 3.20 is acceptable? I guess it's possible that security updates will come in for Jessie's 3.14.5-1+deb8u1 and raise the version a little, but would eg `libgtk-3-0 (<< 3.15)` be safer? 

@hhhorb Do we know exactly at what point breakage starts?

**Comment by ghost** [comment]
3.18 ought to be good. The themes should still work with it albeit maybe not perfectly. 3.15 makes no sense because that's a development version (as is 3.17, 3.19). 3.16 also was still fine iirc.

**Comment by johnraff** [comment]
Ah, so `<< 3.20` makes sense.

**Comment by ju1ius** [comment]
> Ah, so << 3.20 makes sense.

@johnraff The main Bunsen theme was working perfectly for me with 3.18.*

**Comment by ju1ius** [comment]
@johnraff @2ion @hhhorb The gtk-3.20 branch of Greybird [was recently merged into master](https://github.com/shimmerproject/Greybird/commit/688b10c49752e89972d051d8cd67b8004b9db952), so I guess there's something to  build upon now :wink:

**Comment by despin** [comment]
Hi, I wanted to notify you that Greybrid already supports GTK 3.20, and wanted to know if the project is alive, there no commit since August :cry:

**Comment by ghost** [comment]
On Wed, Dec 21, 2016 at 06:55:39PM -0800, Despin wrote:
> Hi, I wanted to notify you that Greybrid already supports GTK 3.20, and wanted to know if the project is alive, there no commit since August :cry: 

Thanks for the notice!  Our theming man @hhhorb is busy with real life
and mostly AFK at this moment, this is why there are no real changes to
the themes.

The base theme also didn't change gradually but underwent structural
changes, so the task is basically to rewrite everything.

Perhaps we should consider just shipping one of the well-supported
vanilla themes, or just keep the default Adwaita (available for both
gtk2 and gtk3)?

**Comment by despin** [comment]
Thank you for confirming the status of the project! I understand the problems, but I was afraid they were dead! Hopefully they can update the thenes. Oh, and they plan to take another version of the system when Stretch comes out?

**Comment by hhhorb** [comment]
I'm back after a long absence. Please see my comment here...
https://github.com/BunsenLabs/bunsen-themes/issues/16#issuecomment-282858021

**Comment by annoyingduck** [comment]
@hhhorb I just discovered these themes from a G+ post.  I'm in love, been looking for dark themes that use a grey highlight color. Obviously some of the GTK3 issues are tough, but it's still very much a usable theme.  I'll be following this project closely and looking forward to future updates. Always available to test - I run Arch XFCE and Manjaro Cinnamon, both GTK-3.22.  I'll check out the Archlabs theme package you linked too.
EDIT: this theme package seems to have "some" better GTK-3.22 integration (text isn't compressed, XFCE panel plugin buttons are better, gnome applications have better window borders) : https://github.com/Dobbie03/ARCHLabs-Themes

**Comment by hhhorb** [comment]
The ARCHLabs themes were built from Numix using the oomox GUI tool, which I couldn't get to build on stretch when I briefly tried it last week...
https://github.com/actionless/oomox

The ARCHLabs themes have some problems as a result, so I spent my time reworking Greybird for stretch...
https://github.com/BunsenLabs/bunsen-themes/issues/16#issuecomment-285122442

I suggest reading the discussion on that issue and continuing the conversation there. Closing so that all discussion is in one place.

-edit- Oh, and thanks for both the compliment and the interest! :D
